collections/src/main/scala/strawman/collection/immutable/ChampHashSet.scala

```diff
@@ -0,0 +1,607 @@
+            deepContentEquality(this.content, node.content, payloadArity, content.length)
+      case _ => false
+    }
+
+  private def deepContentEquality(a1: Array[Any], a2: Array[Any], splitAt: Int, length: Int): Boolean = {
+    if (a1 eq a2) return true
+
+    for (i <- Range(0, length))
+      if (a1(i) != a2(i)) return false
+
+//    // compare local payload
+//    for (i <- Range(0, splitAt)) {
+//      val o1 = a1(i)
+//      val o2 = a2(i)
+//
+//      if (o1 != o2) return false
+//    }
+//
+//    // recursively compare nested nodes
+//    for (i <- Range(splitAt, length)) {
+//      val o1 = a1(i).asInstanceOf[SetNode[A]]
+//      val o2 = a2(i).asInstanceOf[SetNode[A]]
+//
+//      if (o1 != o2) return false
+//    }
+
+    return true
+  }
+
+}
+
+@SerialVersionUID(2L)
+private[this] sealed class HashCollisionSetNode[A](val hash: Int, val content: Vector[A]) extends SetNode[A] {
+
+  import SetNode._
+
+  require(content.size >= 2)
+
+  override def contains(element: A, hash: Int, shift: Int): Boolean =
+    content.contains(element)
+
+  override def updated(element: A, hash: Int, shift: Int, effect: Effect[A]): SetNode[A] =
+    if (this.contains(element, hash, shift)) {
+      this
+    } else {
+      effect.setModified
+      new HashCollisionSetNode[A](hash, content.appended(element))
+    }
+
+  /**
+    * Remove an element from the hash collision node.
+    *
+    * When after deletion only one element remains, we return a bit-mapped indexed node with a
+    * singleton element and a hash-prefix for trie level 0. This node will be then a) either become
+    * the new root, or b) unwrapped and inlined deeper in the trie.
+    */
+  override def removed(element: A, hash: Int, shift: Int, effect: Effect[A]): SetNode[A] =
+    if (!this.contains(element, hash, shift)) {
+      this
+    } else {
+      effect.setModified
+      val updatedContent = content.filterNot(element0 => element0 == element)
+
+      updatedContent.size match {
+        case 0 => SetNode.empty
+        case 1 => new BitmapIndexedSetNode[A]($bitpos($mask(hash, 0)), 0, updatedContent.toArray)
+        case _ => new HashCollisionSetNode[A](hash, updatedContent)
+      }
+    }
+
+  override def hasNodes: Boolean = false
+
+  override def nodeArity: Int = 0
+
+  override def getNode(index: Int): SetNode[A] =
+    throw new IndexOutOfBoundsException("No sub-nodes present in hash-collision leaf node.")
+
+  override def hasPayload: Boolean = true
+
+  override def payloadArity: Int = content.size
+
+  override def getPayload(index: Int): A = content(index)
+
+  override def sizePredicate: Int = SizeMoreThanOne
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case node: HashCollisionSetNode[A] =>
+        (this eq node) ||
+          (this.hash == node.hash) &&
+            (this.content.size == node.content.size) &&
+            (this.content.filterNot(element0 => node.content.contains(element0))).isEmpty
+      case _ => false
+    }
```

Comment:
Where is this used?

Reply:
Are you referring to the `equals` method? It's part of the structural equality implementation and gets invoked in the `ChampHashSet#equals` in the expression `(this.rootNode == set.rootNode)`. So, whenever you compare two sets that have hash collisions, then this method is executed.

Reply:
Yes, the `equals` method. Thanks for the clarification. So why don’t you need to also implement `hashCode` in that case?

Reply:
`BitmapIndexedSetNode` and `HashCollisionSetNode` are private to the file and are never hashed or put into any hashed data structures that's why I omitted and the `hashCode` implementation; these classes represent the trie encoding of the data structure. On the other hand, `equals` is called as shown before to faster compare to trie data structures.

Reply:
Excellent, thanks for the clarification! Maybe you can put a comment in the file about that?

Reply:
Done.